### PR TITLE
[search] [assessment-tool] A tool to merge results from two assessors.

### PR DESCRIPTION
**Pull request by mpimenov**

### CODE REVIEW

File: search/search_quality/assessment_tool/python/merge.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+        raise Exception(f"Tried to merge two non-equivalent samples, line {line_number}")
+
+    useless_key = "useless"
+    useless_a = useless_key in a and a[useless_key]
+    useless_b = useless_key in b and b[useless_key]
+
+    if useless_a and not useless_b:
+        return b
+    if useless_b:
+        return a
```
Comment (tatiana-yan) [inline search/search_quality/assessment_tool/python/merge.py]: может будем вручную решать? или один доразмечает или второй соглашается что useless?

Comment (mpimenov) [reply search/search_quality/assessment_tool/python/merge.py]: Посмотрели на вывод, решили оставить как есть.
Добавил print().

File: search/search_quality/merge_samples.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
+import argparse
+import copy
+import json
+
+
+def read_samples(path):
+    with open(path, "r") as f:
+        lines = f.readlines()
+        return [json.loads(x) for x in lines]
```
Comment (burivuh) [inline search/search_quality/merge_samples.py]: [json.loads(x) for x in f] ?

File: search/search_quality/merge_samples.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+def doubles_equal(a, b):
+    EPS = 1e-6
+    return abs(a - b) < EPS
+
+
+def points_equal(a, b):
+    for coord in ["x", "y"]:
+        if not doubles_equal(a[coord], b[coord]):
+            return False
+    return True
```
Comment (burivuh) [inline search/search_quality/merge_samples.py]: return doubles_equal(a['x'], b['x']) and doubles_equal(a['y'], b['y']) ? It seems less confusing imo

File: search/search_quality/merge_samples.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+def samples_equal(a, b):
+    """
+    Returns whether two samples originate from the same unassessed sample,
+    i.e. found results and uselessness are not taken into account.
+    """
+    for field in ["query", "locale"]:
+        if a[field] != b[field]:
+            return False
+
+    pos_key = "position"
```
Comment (burivuh) [inline search/search_quality/merge_samples.py]:
```
a_pos, b_pos = a.get('position', 0), b.get('position', 0)
if not points_equal(a_pos, b_pos):
    return True
```
?

File: search/search_quality/merge_samples.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+    a_has_pos = pos_key in a and a[pos_key]
+    b_has_pos = pos_key in b and b[pos_key]
+    if a_has_pos != b_has_pos:
+        return False
+    if a_has_pos and not points_equal(a[pos_key], b[pos_key]):
+        return False
+
+    vp_key = "viewport"
```
Comment (burivuh) [inline search/search_quality/merge_samples.py]: I don't think it is necessary not to use this key inline, but ok.

File: search/search_quality/merge_samples.py (hunk 3 above)
Comment (mpimenov) [reply search/search_quality/merge_samples.py]: done

File: search/search_quality/merge_samples.py (hunk 4 above)
Comment (mpimenov) [reply search/search_quality/merge_samples.py]: done

File: search/search_quality/merge_samples.py (hunk 2 above)
Comment (mpimenov) [reply search/search_quality/merge_samples.py]: done

File: search/search_quality/merge_samples.py [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+            if match_to_b[j] >= 0:
+                continue
+            if results_equal(a[i], b[j]):
+                match_to_a[i] = j
+                match_to_b[j] = i
+                break
+    return match_to_a, match_to_b
+
+
+def merge_relevancies(a, b):
```
Comment (burivuh) [inline search/search_quality/merge_samples.py]: doc string at least on the protocol will be much appreciated

File: search/search_quality/merge_samples.py [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+        return b, True
+    return a, False
+
+
+def merge_two_samples(a, b, line_number):
+    if not samples_equal(a, b):
+        raise Exception(f"Tried to merge two non-equivalent samples, line {line_number}")
+
+    useless_key = "useless"
+    useless_a = useless_key in a and a[useless_key]
```
Comment (burivuh) [inline search/search_quality/merge_samples.py]: Can you use get()?

File: search/search_quality/merge_samples.py [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+
+    if useless_a != useless_b:
+        print(line_number, "useless:", useless_a, useless_b)
+
+    if useless_a and not useless_b:
+        return b
+    if useless_b:
+        return a
+
+    # Both are not useless.
```
Comment (burivuh) [inline search/search_quality/merge_samples.py]: May be to another func?

File: search/search_quality/merge_samples.py [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+        rb = b[res_key][j][rel_key]
+        rc, ok = merge_relevancies(ra, rb)
+        res[rel_key] = rc
+        if not ok:
+            print(line_number, ra, rb, a[res_key][i]["name"])
+            pass
+        c[res_key].append(res)
+
+    # Add all unmatched results as is.
+    c[res_key].extend([a[res_key][i] for i in range(len(match_to_a)) if match_to_a[i] < 0])
```
Comment (burivuh) [inline search/search_quality/merge_samples.py]: at least format please

File: search/search_quality/merge_samples.py [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+        if j < 0:
+            continue
+        res = a[res_key][i]
+        ra = a[res_key][i][rel_key]
+        rb = b[res_key][j][rel_key]
+        rc, ok = merge_relevancies(ra, rb)
+        res[rel_key] = rc
+        if not ok:
+            print(line_number, ra, rb, a[res_key][i]["name"])
+            pass
```
Comment (burivuh) [inline search/search_quality/merge_samples.py]: no need in pass

File: search/search_quality/merge_samples.py [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+
+    lst = [x for x in match_to_a if x < 0]
+
+    rel_key = "relevancy"
+
+    c = copy.deepcopy(a)
+    c[res_key] = []
+
+    for i, j in enumerate(match_to_a):
+        j = match_to_a[i]
```
Comment (burivuh) [inline search/search_quality/merge_samples.py]: Why do you need j from enumerate if you're reassigning it right after?

Comment (mpimenov) [reply search/search_quality/merge_samples.py]: That's a leftover. Removed.

File: search/search_quality/merge_samples.py (hunk 10 above)
Comment (mpimenov) [reply search/search_quality/merge_samples.py]: That was a reminder. Turned out we're ok with a simple print here.
done

File: search/search_quality/merge_samples.py (hunk 7 above)
Comment (mpimenov) [reply search/search_quality/merge_samples.py]: done

File: search/search_quality/merge_samples.py (hunk 8 above)
Comment (mpimenov) [reply search/search_quality/merge_samples.py]: I prefer to leave it as is.

File: search/search_quality/merge_samples.py (hunk 9 above)
Comment (mpimenov) [reply search/search_quality/merge_samples.py]: What do you mean?
yapf suggests
```
c[res_key].extend(
        [a[res_key][i] for i in range(len(match_to_a)) if match_to_a[i] < 0])
```

Doesn't look better to me.

File: search/search_quality/merge_samples.py (hunk 6 above)
Comment (mpimenov) [reply search/search_quality/merge_samples.py]: done

File: search/search_quality/merge_samples.py (hunk 9 above)
Comment (burivuh) [reply search/search_quality/merge_samples.py]: By format I mean this
```
c[res_key].extend(
    a[res_key][i]
    for i in range(len(match_to_a))
    if match_to_a[i] < 0
)
```

Comment (mpimenov) [reply search/search_quality/merge_samples.py]: done

**Pull request merged**